### PR TITLE
feat: Add markdown linter for curriculum, with checks of test YAML

### PR DESCRIPTION
**Pull request by ojeytonwilliams**
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

This is a companion to #35735 and also addresses #35729.

@raisedadead suggested we should catch bad YAML syntax before  running the tests.  This is achieved by attempting to parse the YAML code blocks using the same library as `challenge-md-parser` and reporting any errors.  

It also lints the markdown more generally, though with a lot of rules disabled.  This found a few problems that had not been reported (to my knowledge!), so it seems useful.  However, I'm not sure exactly which rules we want to be enforcing, and I don't know exactly where best to put these files, so any input would be great!

### CODE REVIEW

**Comment by raisedadead** [comment]
Hey @ojeytonwilliams can you rebase this PR for QA. Thanks for your patience.

**Comment by ojeytonwilliams** [comment]
@raisedadead No problem.  I'm not sure what (if anything) to do about the other languages, but it seems to be working fine for English.

File: package.json [hunk 1]
```diff
@@ -19,6 +19,7 @@
     "hooks:install": "node node_modules/husky/husky.js install",
     "hooks:uninstall": "node node_modules/husky/husky.js uninstall",
     "lint": "eslint .",
+    "lint:curriculum": "markdownlint -r markdown-yaml.js -c .markdownlintrc.json curriculum/challenges/english/**/*.md",
```
Comment (raisedadead) [inline package.json]: Throw this in a script? https://github.com/freeCodeCamp/freeCodeCamp/tree/master/tools/scripts/ci

**Comment by raisedadead** [comment]
This is a great start, I think we can throw this in a script (a new one or add into an existing one) in the location I have left in my comment in the review and take it from there. This way we are also able to use the `LOCALE` env var to get it working later with any language.

**Comment by ojeytonwilliams** [comment]
@raisedadead I feel daft now, because I must have skimmed your comments as I only just noticed the bit about `tools/scripts/ci`.  Sorry about that!  I'll take a look at that and see if I can merge the two and/or remove redundant stuff.

Also, this is supposed to fail the travis checks, because they now include the new script and there are a ton of challenges that have formatting issues.

**Comment by raisedadead** [comment]
> Also, this is supposed to fail the travis checks, because they now include the new script and there are a ton of challenges that have formatting issues.

Yes, I understand, we might in that case fix up the formatting on this PR? Let's get others to work on this too?

File: package.json [hunk 2]
```diff
@@ -18,7 +18,9 @@
     "format": "npm run lint -- --fix",
     "hooks:install": "node node_modules/husky/husky.js install",
     "hooks:uninstall": "node node_modules/husky/husky.js uninstall",
-    "lint": "eslint .",
+    "lint": "npm-run-all p lint:*",
```
Comment (raisedadead) [inline package.json]:
```suggestion
    "lint": "npm-run-all -p lint:*",
```

**Review by raisedadead — CHANGES_REQUESTED** [review]

**Comment by ojeytonwilliams** [comment]
> Yes, I understand, we might in that case fix up the formatting on this PR? Let's get others to work on this too?

Yeah, we'd have to do it in this since we need newlines around the code fences.  I'd be more than happy if anyone wanted to help fixing the challenge files.

**Comment by ojeytonwilliams** [comment]
It seems that `ensure-challenge-formatting.js` is still useful in that it validates frontmatter.  However parsing the entire markdown seems redundant, given that exactly that happens during testing.  Regardless, this is feeling tangential to me, so maybe it would be better to just fix the linting in this PR

@RandellDawson @moT01 do you think you'll have time to look at fixing the challenge files?  If not, don't worry - just let me know and I'll figure it out.

**Review by ValeraS — CHANGES_REQUESTED** [review]
We can use prettier for formatting and linting md-files. 
Do tests catch invalid yml-blocks?

File: package.json [hunk 3]
```diff
@@ -70,7 +72,7 @@
   },
   "husky": {
     "hooks": {
-      "pre-commit": "lint-staged"
+      "pre-commit": "lint-staged && npm run lint:challenges"
```
[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply package.json]: Good point - I'll sort that out.

**Comment by ojeytonwilliams** [comment]
> We can use prettier for formatting and linting md-files.

I'm not sure we can.  As I understand it, prettier parses the code into an AST and then uses that to create new code.  The problem is that 

````
text

```
code
```
````
generates a different AST from
````
text
```
code
```
````
so prettier will (I think) end up creating different output, rather than flagging up the second one as incorrect.

> Do tests catch invalid yml-blocks?

They do.  However, I haven't written any tests to verify the linter is doing its job. I'll see what I can do.

**Comment by ValeraS** [comment]
> I'm not sure we can. As I understand it, prettier parses the code into an AST and then uses that to create new code. The problem is that

I propose to add the next:
```json
"scripts": {
    "format:other": "npm run prettier -- --write",
    "lint:other": "npm run prettier -- --check",
    "prettier": "prettier \"**/*.{md,css}\""
}
"lint-staged": {
    "*.{md,css}": [
      "prettier --write",
      "git add"
    ]
  }
```
also add `.prettierignore` file
```
node_modules
**/.cache
**/public
client/static
```



> > Do tests catch invalid yml-blocks?
> 
> They do. However, I haven't written any tests to verify the linter is doing its job. I'll see what I can do.

Do you mean current tests or lint checks from this PR?

**Comment by ojeytonwilliams** [comment]
Sure, I'll add the prettier stuff in the next commit.

> > > Do tests catch invalid yml-blocks?
> > 
> > 
> > They do. However, I haven't written any tests to verify the linter is doing its job. I'll see what I can do.
> 
> Do you mean current tests or lint checks from this PR?

Both, actually.  Since https://github.com/freeCodeCamp/freeCodeCamp/pull/35735 the tests catch bad yml properly.  This PR just catches it (and other md issues) earlier.

**Comment by ojeytonwilliams** [comment]
I've added some tests and implemented @ValeraS's suggestions, with a couple of changes.  Firstly, `node_modules` is ignored by default, but it needed to ignore `fixtures` so it wouldn't 'fix' the tests!  Secondly, I went with `--debug-check` for format, since it seems like a good idea to err on the side of caution.

Anyway, adding prettier was a good call.  It cleans up some stuff I didn't expect it to and cuts down the amount of files we need to change manually.

For the moment, I haven't extended the markdown linter over the guide, since I don't know how to automatically fix files.

**Comment by ValeraS** [comment]
@ojeytonwilliams What lint errors do not fix a prettier?
We can check a yaml syntax simply by execute `getChellangesFoLang`.

**Comment by ojeytonwilliams** [comment]
`prettier` misses a bunch of stuff.  If you try running `npm run lint:other` and then `npm run lint:challenges` you can see some of them.

Some of the things `prettier` doesn't reliably catch are
- Use of indentation instead of code fences
- Lack of newline before a code fence
- Multiple `h1` elements
- Skipping header numbers e.g. `h2` -> `h4`

It's particularly bad if a newline doesn't appear before a code fence, as it will try to fix it. For example in `curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-data-structures/-iterate-through-the-keys-of-an-object-with-a-for...in-statement.english.md` if you delete the empty line before the first code fence and `npm run lint:other` you end up with
`````
...this could look like:
```js
for (let user in users) {
  console.log(user);
}

// logs:
Alan
Jeff
Sarah
Ryan

````
`````
The with the closing four backticks wrapping the second fence.

I'm sure `markdownlint` will miss some mistakes, but it at least catches this.

**Comment by ojeytonwilliams** [comment]
One downside of prettier I've just discovered.  You can't write 'bad' code, it'll format it nicely regardless.  This is actually a problem, since at least one challenge demonstrates poorly formatted code to contrast with nicer code.

One PR I'm reviewing has the following [challenge](https://github.com/freeCodeCamp/freeCodeCamp/blob/048c361815d4fd3f564bcfc407f9622072df6d67/curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-multiple-conditional-ternary-operators.english.md).  If we format it with prettier, it stops making sense.

You can see the before and after [here](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBadAdKBLAJkgAgFYAOAdlzICMy4BmANjIBYBGagJio6Na49ZYY2GABs4hAKoBnOAQCyAV1HCADuIIBhaLhHZoAQ1EEAFABU4AJygHLATwCUBAPKqrBmBEvSsYABZG4lAA5nBmdm6EglAAbnhwEJKWooQA5H4wMKrSSAD0udJgltgAtlQGAHSQJblgtXYA6gBSAEJmzKlY6KhYWADEfQQAInCFxarC0FgAPLJgk1AEeAC8qbijRdgT+lCpAHxYAJKLMH5yqpZwcRCK0gT+gQihADQEdjcEt3C4BAYE0tgQhpppB1ntIFBdAsjAQIG5LB4vNNahAwRUCABND5gAyLIzSCD3AKAginOAlUkQUJkyyUolwMAAawIADMvAQSso1BoIVCdtIKlgzGdWRBRKIIAB3QHBVmKKDzHafWR3bAs15wUSyJbq36Qgia7XSGAeckIGB3Tz0pkcrlbHk6PTQHK9KAAAw9ACsfFAWfLFdBWYDcABxS6myzOSwAUQAjoojCYDK8qE5gFgCDrTH9lrmCKmCOnFpmCJcYIprAQMCA-riflRfpcDfGjNWANwZggAX07hrkauzBD2+bTnczZYri2rf2wd2C4fglnbnZ7xb7hbHpbg5cr1Ybs4I87gEeXxdXq49btdwrkBioEBicj9CoW91x+bkl1QkuKmQQyplW0VHtOQQVROBwUdaFjDhdxPG8ZFQQgpBXUvb0sGfANFhZYMw2PRcozjBNRCTFNR2LCdKyTAhc2WEcCAAfirGs9XrRs5DgFtRGrKti0zQhqOHAsmOnJY5wXKweM3Qg9zEw8JKXEAOygC8PVdAAJKUrisV5TgPaQ-BuUR6zkL4fmlU43ybAw7luQDOWA9RQKQyDISdGwYPhREEJRMECAswzFBgAgLlg2lg3NDwlRKAw7A5AxGTkd4K3ucCCACSwfitcNcHRAAxdk4AADwMEonJQqAsDQn1MNfHDITwiNCK40j6KLcdt0nQdaPokSWLrfN2ObYieIEv4hKcPqZ3E-DJJAAgZJAfcZpPJSsFUq9KqgZE5gWA4toGAhjmNSxFCwn1ZgZV8VlSQETrOhZpH2LAZDkBzuWc8DXL5QxPLgrxpEQr6lhOEUwLBfwGUZABlbBgm23yILlF8lStSGbQHP4oEUMorDk1QIABYRH1eKA4GCKLH1hWkAC8rAgQUEd2nZ9v6QYLGNC7mcDG74E556tsvOwSm42BRgtCrM1QUlipgQhwYg9GYbhhGXP+QLjOVN67ScggFe+9yYTC7zAcR-aSxl41oZgYoQkIGzZEsGATFyCoAGoGIwDBpAAKk9jB-d9hj3cDv2UO9v2Q69oOvdD4OPejsPE-jhjcmCbAKj552kIcV5Un1pXYfhoG-IMoyfi+ICPr1lzeUNv6EXg02XNSBxlKlmWirlmuvsLlWTFYAAGBwS6RsvlB+SipxAAmiewR9q3RAA5CB4FJAJgpxVQRCMbAaaiwMYr-bxNyz63beCe3pEd7OziZIuoAH4eaLzVJZ70R9Ujzgu7+V+Gn5HmbdW5ctw7mnu-YmcBF4EBXmvU4Hg3zbxNKIPeB9FhH0XE9NunZpbwC7vLWuv8H4mFQHwQBatx6aynsxMmFNIHQNgXIeBm8DBIN3vvV8GCrA+gtmfG2Mor43xMH3f+pCOBOB6qkWhlM4BfwIPnQhUNiFiPIcDShk9Oq7hANI+hIBl6ryYRvRBO8UEcOih4TBrd24EFwbLAhvciH92HqPYc6jQFdWrHTSwEAGEGPXggreJjUGcIsdw0+4tz4CN+NfKwt8lFOIka-LxEA5EKIcfE-+zigFuOoZ4+mvi4FGMCcg4J5jj5YOUlVNSW0dpXRZq6Q6mgAjiieHIaGcBvgzG5osG6DwWkhDgO074AsZi6BiEsXAqxvSoFkMM1m7ovQ1X9K+ERj9sYlHIutV0qyAHKUvK6ZEYz5kHIKHU6A8zDrQzFEFHYXSzk9MmakAkogbnQBGQst06FfTLKVDs9Z5EOpgNMOsocBBn5MTfoTD+siFrApxnrMFk15E6PnjCtIyTUjKQ2jMU5WE9ggGeCAOEj1kCgFsN4yUAAFWwCBpDIBrDECAeACVLQREybc0NVAGDADKZANtFBwEJRkEWDQ-AiFGFysAgzaUfxEHYelYBr4srurEylCJggxWQCyfEgqQDeiKi0NliUYDQ1KnAAAMoCOAWqdWEv1bDIEcAACKigDE2q1Lqrl3grD0pipYRkuApRQBZRcQEMAGh4FOMgEgg9CWhVkA0BEqh6UXFGFYR8LL4wGLVXCOlKAbKoDJt8b4LLLjxmwJcNVBgNUGHdbIQlsgSjYD5adXVAJHUurdUgbVHrCUmioBG3AUakAcD7QibAKCQjaBKJqlAUBoDWsJV8Mwd4809tkF2LsQA)

We can work around this by using `<pre><code>` blocks, see [here](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBadAdKBLAJkgAgFYAOAdlzICMy4BmANjIBYBGagJio6Na49ZYY2GABs4hAKoBnOAQCyAV1HCADuIIBhaLhHZoAQ1EEAFABU4AJygHLATwCUBAPKqrBmBEvSsYABZG4lAA5nBmdm6EglAAbnhwEJKWooQA5H4wMKrSSAD0udJgltgAtlQGAHSQJblgtXYA6gBSAEJmzKlY6KhYWADEfQQAInCFxarC0FgAPLJgk1AEeAC8qbijRdgT+lCpAHxYAJKLMH5yqpZwcRCK0gT+gQihADQEdjcEt3C4BAYE0tgQhpppB1ntIFBdAsjAQIG5LB4vNNahAwRUCABND5gAyLIzSCD3AKAginOAlUkQUJkyyUolwMAAawIADMvAQSso1BoIVCdtIKlgzGdWRBRKIIAB3QHBVmKKDzHafWR3bAs15wUSyJbq36Qgia7XSGAeckIGB3Tz0pkcrlbHk6PTQHK9KAAAw9ACsfFAWfLFdBWYDcABxS6myzOSwAUQAjoojCYDK8qE5gFgCDrTH9lrmCKmCOnFpmCJcYIprAQMCA-riflRfpcDfGjNWANwZggAX07hrkauzBD2+bTnczZYri2rf2wd2C4fglnbnZ7xb7hbHpbg5cr1Ybs4I87gEeXxdXq49btdwrkBioEBicj9CoW91x+bkl1QkuKmQQyplW0VHtOQQVROBwUdaFjDhdxPG8ZFQQgpBXUvb0sGfANFhZYMw2PRcozjBNRCTFNR2LCdKyTAhc2WEcCAAfirGs9XrRs5DgFtRGrKti0zQhqOHAsmOnJY5wXKweM3Qg9zEw8JKXEAOygC8PVdAAJKUrisV5TgPaQ-BuUR6zkL4fmlU43ybAw7luQDOWA9RQKQyDISdGwYPhREEJRMECAswzFBgAgLlg2lg3NDwlRKAw7A5AxGTkd4K3ucCCACSwfitcNcHRAAxdk4AADwMEonJQqAZguCCwLBDD-VfHDITwiNCK40j6KLcdt0nQdaPokSWLrfN2ObYieIEv4hKcQaZ3E-DJJAAgZJAfd5pPJSsFXRDwL2ZFqoOSqoGROYFkO-pBmOY1LEULCfVmBlXxWVJAWu26FmkfYsBkOQHO5Zzdt5dyYTC7zpB2vySTJAhaog-wGUZABlbBgmO3yILlF8lSteGbQHP4oEUMorDk1QIABYRH1eKA4GCKLH1hWkAC8rAgQU0dOnZzqgAYCAsY17s5wNnvgAWvqOy87BKbjYFGC0KszVBSWKmBCFh8EziZZHUYhjGDKMn4viA-6YZcoHoNhLz4PB9HDpLZXjURmBihCQgbNkSwYBMXIKgAagYjAMGkAAqAOMDDkOGL9iPQ5QoPQ+jwPI8DmOo-9pPY4ztOGNyYJsAqUWvaQhxXlSdXcaRlG0Zc-5AuM5VfrtJzTcBqCdhBq2vBtlzUgcZTFeVorVZbsEK+1qATFYAAGBxdeHfXlB+SipxAMmKewR9q3RAA5CB4FJAJgpxVQRCMbAmaiwMYr-bxN0Lp2XeCN3pA9ovNcr1HJ5nmi81SNe9EfKkUu5d37jy-rPW2tcDZbh3Cvf+lM4BbwILvfepwPBvhPiaUQ59L6LGvouT6fdOxK3gEPNWZtQFVxMKgPgECa4L3rsvZiNM6YIKQSguQaCj4GEwWfC+r58FWB9Pbe+zsZTP1fiYMeVCaEcCcP1VILD6ZwCAQQMuFCEZgNkXQ3aUDF4wN6tWJRbCQA7z3pww+GDT7YP4dFDwBDe79wICQlW5DAaUM-jPOeejGE9V3CAFmlgIDsPMQfdBx9rE4IEfYoRd85YP3Eb8F+Vg36aKod-BRgSICqPUe4tJnidF+QYUvPxK8skhNQZYiJWCol2JvoQ5SWBLyuhOo9LmrpeaaACOKJ4chEZwG+DMIWixnoPB6SEOA-TvjixmLoGISxcCrG9KgWQ0zuZoR9JhV80jP6ExKORLarodkT2nkQ90akjrIjmdzFpBQ2nQG5rzRGYogo7CGfckZizUgElEK86AMzzlunQr6BqSpjkmD2eRbqsDTB7KHAQb+TE-7kwASo5asKiYwwRTNNRxiN5orSFk1IylVJXg5h8vYIBnggDhB9ZAoBbBBMlAABVsAgaQyAawxAgHgKlq0ERMm3IjVQBgwAymQM7RQcBqUZGlg0PwIhRgirAJM9lACRB2E5WAF+fLXopOZQiYIMVkAsnxNKkA3oiotAFYlGAiNSpwAADKAjgCas11LLXIyBHAAAioocxbqtTmpFd4KwnKYqWEZLgKUUA+UXEBDABoeBTjIBIFPaloVZANARKoTl1UPaPj5fGcxBq4QcpQDZVANNvjfD5ZceM2BLgGoMEagwgbZDUtkCUbAEqbrmoBN6v1AakCmqDdSk0VAk24BTUgDg46ETYGwSEbQJRjUoCgNAV11KvhmDvOW0dsguxdiAA), which prettier won't touch, but only if we know that's needed.

@raisedadead if it's just this challenge that does this, we're probably okay, but I don't know if that's the case.

**Comment by ojeytonwilliams** [comment]
I'm not 100% happy with the organisation of this, so any ideas are welcome.  The reasoning behind the two sets of linting rules is simply that the markdown in the guide is much looser than  in the curriculum.  If we lint new files in staging we can make sure new articles are nice and neat, but applying strict rules to the guide without some way of automating it is rather daunting.

The rules that it does enforce on the guide should be relatively easy to apply via regex.

I'll fix the conflict in a sec.

**Comment by raisedadead** [comment]
> [...] This is actually a problem, since at least one challenge demonstrates poorly formatted code to contrast with nicer code [...]

&&

> [...] If we format it with prettier, it stops making sense [...]

Can we have special escaping done somehow, like ESLint allows disabling rules per line, per file, etc.

> [...] The reasoning behind the two sets of linting rules is simply that the markdown in the guide is much looser than in the curriculum. [...]

I can understand your frustration with this. But the real real solution to this would be to have a dedicated contributor editor (something we have been discussing and working on). That way we can cleanup the guide eventually. In all honesty the efforts spent on fixing formatting on Guides to be relaxed is fine, as long as we can improve the flow to make quick edits.

For now this is fine. Content editing is hard enough and doing it with Git is worse.

**Comment by ojeytonwilliams** [comment]
> Can we have special escaping done somehow, like ESLint allows disabling rules per line, per file, etc.

We can and I should have thought of that!  Putting `<!-- prettier-ignore -->` before a code block will make `prettier` leave it alone.  

However, the real problem is that we don't know which challenges need it.  Since we talked about this last I've realised that it's not just deliberately bad code that gets hit; it's anything that has slightly non-standard formatting.  Here are some examples

Before:
ObjectPropsBefore
After:
ObjectPropsPrettier
Before:
MultiDimBefore
After:
MultiDimAfter

These aren't terrible, it's just they no-longer fully reflect the author's intent.  Also a lot of challenge seeds have extra empty lines for users to enter their code, which will get stripped out by `prettier`.

On balance, I don't think we should be using `prettier` on the guide or the challenges.

> a dedicated contributor editor [...] That way we can cleanup the guide eventually. In all honesty the efforts spent on fixing formatting on Guides to be relaxed is fine [...]

An editor would be amazing!  Just to reiterate, I'm absolutely in favour of applying as strict linting as we can to future guide PRs.  That's why I added the 'full' linter to `lint-staged`, I just don't think it should be applied retroactively.

**Comment by raisedadead** [comment]
> However, the real problem is that we don't know which challenges need it.

I think we can do an audit during the QA phase on beta.

> On balance, I don't think we should be using `prettier` on the guide or the challenges.

I am inclined to that too. We want any linting and tooling to augment our work and not make it worse. I think your original approach is just as fine and better than anything.

**Comment by raisedadead** [comment]
> I just don't think it should be applied retroactively.

I agree, and I think we should have some rules and not 'strict rules' as long as it's not platform code. It will be a way better experience than now.

**Comment by ojeytonwilliams** [comment]
Okay, cool.  I'll stop using `prettier` on markdown files (css seems safe) and get to work on the markdown itself so it can pass.  I think @RandellDawson's already sorted out the problems with prism languages, so we're well on the way.

**Review by raisedadead — APPROVED** [review]

**Pull request merged**